src/stream/gzip.rs

```diff
@@ -167,3 +167,167 @@ fn get_header(level: Compression) -> Bytes {
 
     Bytes::from(header)
 }
+
+#[derive(Debug)]
+enum DeState {
+    ReadingHeader,
+    Reading,
+    Writing,
+    ReadingFooter,
+    Done,
+    Invalid,
+}
+
+#[unsafe_project(Unpin)]
+pub struct DecompressedGzipStream<S: Stream<Item = Result<Bytes>>> {
+    #[pin]
+    inner: S,
+    state: DeState,
+    input: Bytes,
+    output: BytesMut,
+    crc: Crc,
+    decompress: Decompress,
+}
+
+impl<S: Stream<Item = Result<Bytes>>> Stream for DecompressedGzipStream<S> {
+    type Item = Result<Bytes>;
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Result<Bytes>>> {
+        let mut this = self.project();
+
+        fn decompress(
+            decompress: &mut Decompress,
+            input: &mut Bytes,
+            output: &mut BytesMut,
+            crc: &mut Crc,
+            flush: FlushDecompress,
+        ) -> Result<(Status, Bytes)> {
+            const OUTPUT_BUFFER_SIZE: usize = 8_000;
+
+            if output.len() < OUTPUT_BUFFER_SIZE {
+                output.resize(OUTPUT_BUFFER_SIZE, 0);
+            }
+
+            let (prior_in, prior_out) = (decompress.total_in(), decompress.total_out());
+            let status = decompress.decompress(input, output, flush)?;
+            let input_len = decompress.total_in() - prior_in;
+            let output_len = decompress.total_out() - prior_out;
+
+            crc.update(&output[0..output_len as usize]);
+            input.advance(input_len as usize);
+            Ok((status, output.split_to(output_len as usize).freeze()))
+        }
+
+        #[allow(clippy::never_loop)] // https://github.com/rust-lang/rust-clippy/issues/4058
+        loop {
+            break match mem::replace(this.state, DeState::Invalid) {
+                DeState::ReadingHeader => {
+                    *this.state = match ready!(this.inner.as_mut().poll_next(cx)) {
```

Comment:
`ready!` hides an early `return Poll::Pending`, if the inner stream is not ready then this will return and leave the current state as `DeState::Invalid`. You should have a `*this.state = DeState::ReadingHeader;` before the match so that if it hits that early return it will go back to the current state (I was pretty undecided in #11 whether it was better to use `ready!` or to explicitly handle the `Poll` in match since this is kinda subtle).